Prototype buildpack that performs vuls scan

ohsh6o opened the pull request:
This is a prototype buildpack that will send scan data to a [vuls](https://vuls.io/) server for processing vulnerable dependencies.

aj-stein-gsa committed:
Add placeholder files.

aj-stein-gsa committed:
Prevent use of obsolete decorate method.

aj-stein-gsa committed:
Create placeholder build script to get external assets.

Reviewing other buildpacks, we ought to retrieve binary
assets needed for our buildpack, such as Github or S3,
and not store them directly inline within the git repo, as
this seems to be best practice.

aj-stein-gsa committed:
Add placeholder compile step script.

aj-stein-gsa committed:
Set up supply script.

aj-stein-gsa committed:
Forgot to fix one old dir reference.

aj-stein-gsa committed:
Test execution of script from profile.d.

aj-stein-gsa committed:
Add warning for detect lifecycle.

As part of mutli-buildpack solution, this cannot be used in the
detect phase of the lifecycle, force exit.

aj-stein-gsa committed:
Drop a test file to confirm execution.

ohsh6o committed:
Merge pull request #1 from flexion/tasks/setup-cf-buildpack

Initial Buildpack for Scanning with vuls

aj-stein-gsa committed:
Mark lifecycle scripts executable.

aj-stein-gsa committed:
Add placeholder config.toml file.

aj-stein-gsa committed:
Correct path for binary extract.

aj-stein-gsa committed:
Remove tarball once binary extracted.

aj-stein-gsa committed:
Add env var debugging.

aj-stein-gsa committed:
Debug env with file.

Somehow the supply lifecycle does not allow me to wrap
`$(env)` correctly in a print statement.

aj-stein-gsa committed:
Need to debug in BUILD_DIR.

aj-stein-gsa committed:
List out path and file organization.

aj-stein-gsa committed:
Use deps_dir, not build_dir for vuls assets.

aj-stein-gsa committed:
Forgot index in path interpolation.

aj-stein-gsa committed:
Better debugging info.

aj-stein-gsa committed:
Remove extraneous depscan sub-dir.

aj-stein-gsa committed:
Fix path in run-scan.sh script.

aj-stein-gsa committed:
Use paths and do not hard-code binary too.

aj-stein-gsa committed:
Run scan and post contents via curl.

aj-stein-gsa committed:
Need to properly escape multi-line shell command.

aj-stein-gsa committed:
Test fuller variable parameterization.

aj-stein-gsa committed:
Fix missed escaping of $ for vars.

aj-stein-gsa committed:
Fix obvious var rename error.

aj-stein-gsa committed:
Woops, do not escape $ bash var sigil.

aj-stein-gsa committed:
Fix some ordering errors and obvious typos.

aj-stein-gsa committed:
Use single quotes in UPLOAD_OPTS.

If not, with nested " quotes in parameterized shell scripts, everything
after "application/json" in the middle leads to subsequent string parts
getting dropped.

aj-stein-gsa committed:
Turn off debugging commands, just execute.

aj-stein-gsa committed:
Fix overwritten rewrite for buildpack idx path.

Accidentally copy-pasted over this during debugging code changes
against cloud.gov staging environment.

aj-stein-gsa committed:
One more var cleanup.

mike-stern commented on bin/supply:
```diff
@@ -0,0 +1,37 @@
+cp -r $ROOT_DIR/lib/dist $VULS_DIR/
+
+echo "Download vuls asset from ${VULS_URL}"
+curl -L -o $VULS_DIR/vuls.tar.gz $VULS_URL
```
pull `vuls.tar.gz` into a variable as it is repeated on L25 & L27 below

ohsh6o replied:
Sounds like a good plan, will do.

ohsh6o replied:
Added.

mike-stern commented on bin/supply:
```diff
@@ -0,0 +1,37 @@
+DEPS_DIR=$3
+INDEX=$4
+
+VULS_URL="${VULS_URL:-https://github.com/future-architect/vuls/releases/download/v0.9.6/vuls_0.9.6_linux_amd64.tar.gz}"
```
make `0.9.6` an externalizable variable at some point in the future

ohsh6o replied:
So the point of the `VULS_URL="${VULS_URL:-https://sensible-default/path/to/download}` is actually to externalize the whole URL. I found out from experimenting with using the `latest` tag that this does not work well. One of my issues now is references another variable to the right of the `:-` and it being weird, but I can definitely try that in a bit.

ohsh6o replied:
I parameterized it anyway.

mike-stern commented on bin/detect:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+#!/usr/bin/env bash
+
+echo "-----> VulsBuildpack/detect"
```
at some point does this become a heuristic scan of deploy targe for known types or does the buildpack plan to be forever explicitly invoked?

ohsh6o replied:
I had read some recommendations that for meta-buildpack and multi-buildpack style that is opt-in only, advisable to code it that way. That being said, are you asking if we should build-in a detection pass that recursively looks in the staged buildpack files for a support dep file?

Before, I would have thought this possible. It was your link to the Brett Mogilefsky presentation at the CF2018 Summit and our follow-up discussion Tuesday that makes me think that might not work. In short: our buildpack has to exist before the others do, and stages a hook file into others later to be seen, but it does not see at the detect lifecycle phase when `bin/detect` is called any of the Python files from our 10x-dux-app for example, and especially the `Pipfile` it needs to look for.

ohsh6o replied:
Also, this review comment was useful in a unrelated way: you clued me into the double shebang.

ohsh6o replied:
I am not going to close this yet, but I am opening up PR for review. I think we can wait for Bret's viewpoint and decide if we can do this, and prioritize.

aj-stein-gsa committed:
Remove double shebang.

Thanks to @mike-stern for surfacing this during an unrelated review
comment all the same!

aj-stein-gsa committed:
Parameterize vuls version in supply script.

Make it easier to update while keeping the Github release download URL
consistent.

aj-stein-gsa committed:
Parameterize download path of tarball.

Follow PR feedback to keep tarball path as variable.

aj-stein-gsa committed:
Workaround for quote escape and script executable.

Working around issues with quote escape in default parameters. No matter
how I quote the default parameter in the parameter substitution causes
problems no matter what I do.

UPLOAD_OPTS="${UPLOAD_OPTS:--X POST -H "Content-Type: application/json" -d @${RESULTS_DIR}/current/localhost.json ${VULS_HTTP_SERVER}}"

UPLOAD_OPTS="${UPLOAD_OPTS:--X POST -H 'Content-Type: application/json' -d @${RESULTS_DIR}/current/localhost.json ${VULS_HTTP_SERVER}}"

UPLOAD_OPTS="${UPLOAD_OPTS:--X POST -H \'Content-Type: application/json\' -d @${RESULTS_DIR}/current/localhost.json ${VULS_HTTP_SERVER}}"

UPLOAD_OPTS="${UPLOAD_OPTS:--X POST -H \"Content-Type: application/json\" -d @${RESULTS_DIR}/current/localhost.json ${VULS_HTTP_SERVER}}"

None of these prevent from bash breaking the string for the header
Content-Type and its value into a separate string, as bash indicates
in debugger mode.

+ /usr/bin/curl '"-X"' POST -H '"Content-Type:' 'application/json"' -d @/home/vcap/tmp/current/localhost.json http://1.2.3.4:5515/vuls

mike-stern reviewed (approved):
As it stands this is ready for merge. The comment on 'detect' is an open question, but not something to hold the merge up.

ohsh6o commented:
> As it stands this is ready for merge. The comment on 'detect' is an open question, but not something to hold the merge up.

Sorry for the delayed response. I will merge this for now. We will review the detect option at a later time.
